### DISCUSSION TITLE
Physics Plugin and Supervisor

### DISCUSSION THREAD

**Discussion by annaa95** [cyberbotics/webots#3953] [Q&A]
Hi, I have a supervised robot which is responsible for removal and creation of another robot at every reset of the simulation.
I also have a plysics plugin to constrain the created robot to 2D motions.
It looks like the plugin does not not run if istantiated by the supervisor. It runs properly only if already there when the world is loaded. 
Do you have some suggestions for achieving the desired behavior?
Should I act on the physics from the supervisor controller instead of using the plugin?
Thanks

**Comment by ad-daniel** [comment]
How do you constraint the motions to 2d from the physics plugin?
Whenever you remove/add a new robot, do you update/get references to the parts you intend to modify? (bodyIds, jointIds, that sort of thing)
If the removal occurs after the simulation has started, you must get these Ids in the `webots_physics_step`, the `webots_physics_init` is only called once at the beginning.

**Reply by annaa95** [reply]
Thanks @ad-daniel for your reply. 
So in my `webots_physics_init` function I createted a 2DPlane and attached to the robot body, as follows:
```
void webots_physics_init() {
  pthread_mutex_init(&mutex, NULL); 
  robot_geom = dWebotsGetGeomFromDEF("ROBOT");
  if (robot_geom)
      robot_body = dGeomGetBody(dSpaceGetGeom((dSpaceID)robot_geom, 0));
   dWorldID world = dBodyGetWorld(robot_body);
   dJointID planeJointID = dJointCreatePlane2D(world, 0);
   dJointAttach(planeJointID, robot_body, 0);
}
```
Since ROBOT does not exist when the simulation starts, but it is inserted in the scene by the Supervisor, the simulation crashes.
So I tried to modify the WorldInfo.physics field from supervisor to add the physics plugin only after Robot creation, but it doesn t work of course.  
I'll try to move the robot related part ot the `webots_physics_step` function as you suggested
